### ISSUE TITLE
Capacitor support

### ISSUE THREAD

**Issue by dylanvdmerwe** [transistorsoft/cordova-background-geolocation-lt#937]
I just wanted to inquire if on your roadmap there is any chance that [Capacitor](https://blog.ionicframework.com/announcing-capacitor-1-0/) will be supported by this plugin in the future?

**Comment by christocracy** [comment]
Why do you want to use this plugin with Capacitor?  Are you building a PWA app?

**Comment by dylanvdmerwe** [comment]
Hi @christocracy 

No not from a PWA perspective. Capacitor will be Ionic's preferred way of integrating with native platforms ([iOS and Android](https://capacitor.ionicframework.com/docs/getting-started/dependencies#ios-development)). As such, this will be a replacement for the cordova layer. 

I would then assume there is a Capacitor version of your plugin (for [iOS ](https://capacitor.ionicframework.com/docs/ios/custom-code/)and [Android](https://capacitor.ionicframework.com/docs/android/custom-code/)) using the exact native code you have without the cordova layer.

**Comment by christocracy** [comment]
This plugin is a pure Cordova plugin.  It cannot be used *without* Cordova.

**Comment by dylanvdmerwe** [comment]
Sorry let me clarify. I understand that _this_ plugin is for Cordova only. I posted this query on this repo because I know many people (and myself included) have used this plugin in their Cordova projects - but as Capacitor is out we are planning a migration path from Cordova to Capacitor and this is a checkbox on the migration list. 

As such, I was querying if a `capacitor-background-geolocation` plugin is on your horizon at all that exposes the functionality we love from the Cordova one?  :smile:

**Comment by christocracy** [comment]
> if a capacitor-background-geolocation plugin

No.  Use plain old Ionic.

**Comment by christocracy** [comment]
Capacitor is for people making PWA apps, apps that run purely in a webview with no need to be released through app stores.  These apps have no access to native APIs.

BackgroundGeolocation, of course, makes heavy use of native apis (Location, Geofencing, Accelerometer, Gyroscope, Magnetometer, SQLite, HTTP).

Capacitor *contains* Ionic.  You need only stick to using pure Ionic.

**Comment by jcesarmobile** [comment]
Just to clarify, Capacitor is a Cordova alternative, and can be used with and without Ionic. It's not PWA only, it can be used to create native and hybrid apps. It has its own set of native plugins and also has Cordova plugin support (not all of them work, but most).

I have not tried your plugin with Capacitor, but if it doesn't work and somebody report the problem with details, we will try our best to make it compatible. In case it's not possible on our end, then the only way might be to have a separate capacitor plugin.

**Comment by christocracy** [comment]
@dylanvdmerwe Just stick with pure Ionic for now.

**Comment by christocracy** [comment]
I created a Hello World Cap app and added `cordova-background-geolocation-lt`.  It works for both iOS and Android.

**Android**
![](https://dl.dropbox.com/s/vzx7kml7yjb6697/Screenshot%202019-06-05%2009.25.06.png?dl=1)

**iOS**
![](https://dl.dropbox.com/s/pb8sjhgduwuiicn/Screenshot%202019-06-05%2009.27.54.png?dl=1)

**Comment by christocracy** [comment]
**Android Setup**:
- Open the root `android/build.gradle`:
```diff
allprojects {
    repositories {
+       maven {
+           url "$rootDir/../node_modules/cordova-background-geolocation/src/android/libs"
+       }
+       maven {
+           url "$rootDir/../node_modules/cordova-plugin-background-fetch/src/android/libs"
+       }
        google()
        jcenter()
    }
}
```
**iOS**
- Open `Info.plist`
- Add the required Usage Strings:
  - *Privacy - Location Always and When In Use Usage Description*
  - *Privacy - Motion Usage Description*
![](https://dl.dropbox.com/s/s3hynskfzhp8nwk/Screenshot%202019-06-05%2009.37.19.png?dl=1)

**Comment by christocracy** [comment]
The manual setup instructions are similar to React Native's manual setup instructions for [Android](https://github.com/transistorsoft/react-native-background-geolocation/blob/master/help/INSTALL-ANDROID.md#open_file_folder-androidbuildgradle) and [iOS](https://github.com/transistorsoft/react-native-background-geolocation/blob/master/help/INSTALL-IOS.md#configure-background-capabilities)

**Comment by christocracy** [comment]
I'm skeptical of the long term for Ionic (and Capacitor):  [Flutter vs Ionic vs React Native](https://trends.google.com/trends/explore?cat=13&date=today%205-y&q=flutter,react%20native,ionic,nativescript)
- Interest in Ionic has been trending down since 2018.
- Interest in [Flutter](https://flutter.io) is now at an inflection point to overtake Ionic, just as React Native did in 2017 (See [flutter_background_geolocation](https://github.com/transistorsoft/flutter_background_geolocation))
- React Native is currently the King.
- I *really* like Flutter.

**Comment by mlynch** [comment]
@christocracy wouldn't count us out yet. Until this Spring you could only use Ionic with Angular, which was a big limiting factor in our growth. We're about to roll out React and Vue support soon as well.

In fact, our most recent community numbers are higher than they've ever been now that we're starting to expand into those other communities, and now that Capacitor is here to improve the Cordova experience that pushed a lot of people away.

Ionic brings some fundamental advantages that React Native and Flutter don't, namely a pure web development experience that also works on mobile, desktop, and web. While Flutter claims to do the latter, Flutter for Web is more like Flash than like a true web development framework, so I wouldn't even count that as true web support.

I imagine the desktop and web stuff doesn't matter to you but hope it helps frame things.

We appreciate you supporting the Cordova side of many an Ionic app and hope you continue to do so!

**Comment by christocracy** [comment]
@mlynch I'll keep my eye on it.  I've looked at the [Capacitor Plugin API](https://capacitor.ionicframework.com/docs/plugins/) this morning and generated a HelloWorld plugin.  I get it.  I can see it'll take a couple of weeks to port a pure Capacitor version of `cordova-background-geoleolocation` and `cordova-plugin-background-fetch`, which would be preferable since it'll have better dependency control, particularly for Android.

This isn't the right place to mention this, but I hope Capacitor will modify its application template to append an `ext` block for global config parameters like React Native does.  This makes life much easier for plugin developers solving Android dependency conflicts.

`android/build.gradle`:
```gradle
buildscript {
    ext {   // <-- globally defined var for use by plugin devs and user's app.
        buildToolsVersion = "28.0.3" 
        minSdkVersion = 16
        compileSdkVersion = 28
        targetSdkVersion = 28
        supportLibVersion = "28.0.0"
    }
    .
    .
    .
}
```
- `android/app/build.gradle`
```
android {
    // Please do not hard-code these:  fetch from ext instead.
    compileSdkVersion rootProject.ext.compileSdkVersion 
    buildToolsVersion rootProject.ext.buildToolsVersion

    defaultConfig {        
        minSdkVersion rootProject.ext.minSdkVersion
        targetSdkVersion rootProject.ext.targetSdkVersion
    }
    .
    .
    .
```

**Comment by mlynch** [comment]
@christocracy thanks for taking a look and for the good feedback. Will look into this with @jcesarmobile and see what we can do!

**Comment by christocracy** [comment]
@mlynch And I hope Capacitor has a plan for supporting AndroidX.

**Comment by bhandaribhumin** [comment]
@christocracy can you please share with me so i can integrate and research on it.

**Comment by christocracy** [comment]
@bhandaribhumin Capacitor support exists in the private repo.  If you have purchased a license key, you should have access to the private repo.  Talk to the person who ordered the product:  they can provide you access to the private repo via the Product Dashboard.

It seems you do not have access to the private.  If you don't, the following url will be 404 for you:
https://github.com/transistorsoft/cordova-background-geolocation

**Comment by bhandaribhumin** [comment]
@christocracy thanks for your advice I’m going to talk my admin regarding this also I need your help for usage?

**Comment by christocracy** [comment]
> Same usage like ionic3?

No.  The plugin has had [Typescript support](https://github.com/transistorsoft/cordova-background-geolocation-lt#ionic-2-with-typescript) for ages.

**Comment by pbramel** [comment]
@christocracy Any idea when the Capacitor functionality will be released publicly? We are working with the private build and after a few hiccups, it seems to be functioning.

**Comment by christocracy** [comment]
> Any idea when the Capacitor functionality will be released publicly?

Typically every 3 months.

**Comment by christocracy** [comment]
Just use the private repo.

**Comment by bhandaribhumin** [comment]
Getting error **Cannot find module 'cordova-background-geolocation'**

![Uploading Screen Shot 2019-09-16 at 3.58.52 PM.png…]()

**Comment by christocracy** [comment]
@bhandaribhumin If you get that error while using the *private* repo, post an issue in the *private* repo.

**Comment by bhandaribhumin** [comment]
@christocracy okay thanks

**Comment by christocracy** [comment]
Capacitor support is published to [3.2.2](https://github.com/transistorsoft/cordova-background-geolocation-lt/pull/996).  See the README for setup instructions.
